### ISSUE TITLE
Does kimchi have to run on the KVM host?

### ISSUE THREAD

**Issue by philled** [kimchi-project/kimchi#952]
I have been having massive difficulties getting kimchi running on my Ubuntu 16.04 KVM host. I end up in dependency hell and it's so bad I'm probably going to wipe the host and re-install from the O/S up.

Is it possible to run kimchi on a guest virtual machine in the host, or does it have to be running on the KVM host itself? I could create a VM and try to get kimchi running on that and it wouldn't matter how many times I have to restart from scratch,

**Comment by pvital** [comment]
@philled you need set nested virtualization on the host machine and then you can use Kimchi in the VM, as a second host. 

Just pay attention about the virtual network in the VM to get it running you need change some ranges. See how to do this here: https://pvital.wordpress.com/2015/09/18/nested-virtualization-and-the-network-issue/

**Comment by jamesfarrugia** [comment]
To follow up on this question, is it possible to configure that kimchi so that it connects to my KVM host over an SSH tunnel?  I have managed to install and run kimchi in a VM while also accessing it from another VM.  I would prefer this setup since I wouldn't like too many dependencies, including X11 ones, installed on my KVM host.

What I need is a way to configure kimchi to manage a KVM host over an SSH tunnel rather than the local one, is this possible please?

**Comment by alinefm** [comment]
Hi @jamesfarrugia 

Kimchi connects to the libvirt service running in the KVM host. If you configure libvirt to use SSH tunel, Kimchi will use it.

To do the libvirt configuration, you will need to edit uri_default value in /etc/libvirt/libvirt.conf according to the links below:

http://libvirt.org/guide/html/Application_Development_Guide-Architecture-Remote_URIs.html
https://libvirt.org/drvqemu.html#uris

**Comment by alinefm** [comment]
Hi @philled 

Any update on that? Could you get Kimchi running on your KVM guest with nested virtualization enabled?

**Comment by philled** [comment]
On 9 Sep 2016 00:59, "Aline Manera" <notifications@github.com> wrote:
>
> Hi @philled
>
> Any update on that? Could you get Kimchi running on your KVM guest with
nested virtualization enabled?
>

Hi there, I'm afraid I gave up on Kimchi. I found the installation process
too complicated and it didn't have clear documentation. On more than one
occasion I ended up with such a messed up state of packages I had to
reinstall the whole system from the OS up.

**Comment by alinefm** [comment]
Hi @philled 

I'd like to understand how complicated was to install Kimchi on Ubuntu 16.04.
I have a virtual machine setup with the same environment and I got no issues to have it up and running.

**Comment by philled** [comment]
Hi Alice,  thanks for letting me know. Could you please send me the link to
the instructions you followed and I'll try it again.

Thanks,
Phill

**Comment by philled** [comment]
Having a closer look at this. Aline, I think you may have installed Kimchi by downloading the source and compiling etc, am I correct? My attempts at installing were to install Kimchi from Ubuntu packages. That's where the dependency hell came in and is why i was asking you which guide you followed. But if you're following the README I presume you're installing it from source not packages.

I'll try to install from source over the next few days and see how I get on.

**Comment by alinefm** [comment]
Hi @philled 

You are correct. I tested it from source code. But the latest packages available at http://kimchi-project.github.io/kimchi/ also work fine on Ubuntu 16.04.
By the end of this week (Sept 30), 2.3 release will be announced and new packages available for download at the same github.io page.
The 2.3 release was fully validated on Ubuntu 16.04 so it must work as well. The test matrix for Ubuntu 16.04 can be found at https://github.com/kimchi-project/kimchi/wiki/Testing-2.3

**Comment by philled** [comment]
>
> You are correct. I tested it from source code. But the latest packages
> available at http://kimchi-project.github.io/kimchi/ also work fine on
> Ubuntu 16.04.
> By the end of this week (Sept 30), 2.3 release will be announced and new
> packages available for download at the same github.io page.
> The 2.3 release was fully validated on Ubuntu 16.04 so it must work as
> well. The test matrix for Ubuntu 16.04 can be found at
> https://github.com/kimchi-project/kimchi/wiki/Testing-2.3
>
Hi, just reporting back on how I got on with this on a new Ubuntu 16.04 VM
(running on KVM).


   1. I installed the Wok dependency packages listed on
   https://github.com/kimchi-project/kimchi/blob/master/docs/ubuntu-deps.md
   under Build Dependencies and Runtime Dependencies.
   But I didn't install the packages listed under Packages required for UI
   development or Packages required for tests.

   Before installing Ginger BAse & Wok I had to install a couple of missing
   dependencies - python-apt and python-imaging.

   2. I then installed the Ginger Base package I downloaded from
   http://kimchi-project.github.io/gingerbase/downloads/

   3. I then installed the Wok package I downloaded from
   http://kimchi-project.github.io/wok/downloads/

   4. I then installed the Kimchi package I downloaded from
   http://kimchi-project.github.io/kimchi/downloads/

   5. Then things got tricky as the screenshots of what you're supposed to
   see for a Wok Login Screen and Kimchi Guest View are broken in
   http://kimchi-project.github.io/kimchi/#build-and-install. Also, I don't
   have X installed on the Kimchi VM so there's no browser to run so I tested
   connecting to localhost 8001 with telnet which worked.
   I then tested connecting from a different PC with a browser and after
   logging in was presented with a Wok screen called "User Activity Log"
   (screenshot below). Is this Kimchi? I don't think it is. Can someone please
   advise what I do from here. The build-and-install page says "To create a
   new guest, click on the "+" button in the upper right corner" - but there
   isn't one.
   [image: Inline images 1].


Regards,
Phill

**Comment by philled** [comment]
>
>
>    1.
>    2. Then things got tricky as the screenshots of what you're supposed
>    to see for a Wok Login Screen and Kimchi Guest View are broken in
>    http://kimchi-project.github.io/kimchi/#build-and-install
>    <http://kimchi-project.github.io/kimchi/#build-and-install>. Also, I
>    don't have X installed on the Kimchi VM so there's no browser to run so I
>    tested connecting to localhost 8001 with telnet which worked.
>    I then tested connecting from a different PC with a browser and after
>    logging in was presented with a Wok screen called "User Activity Log"
>    (screenshot below). Is this Kimchi? I don't think it is. Can someone please
>    advise what I do from here. The build-and-install page says "To create a
>    new guest, click on the "+" button in the upper right corner" - but there
>    isn't one.
>
>
BTW, after encountering the issue above I then rebooted the Kimchi VM and
thyings are now in a worse state because when I go to
https://192.168.0.109:8001/ I get an error message in the browser saying
"502 Bad Gateway     nginx/1.10.0 (Ubuntu)".

Aline, I'm certainly not experiencing the same success you appeared to have
by following the instructions at the Kimchi web site. Hopefully someone can
put me straight here.

**Comment by philled** [comment]
>
> BTW, after encountering the issue above I then rebooted the Kimchi VM and
> thyings are now in a worse state because when I go to
> https://192.168.0.109:8001/ I get an error message in the browser saying
> "502 Bad Gateway     nginx/1.10.0 (Ubuntu)".
>

More info... it looks like nginx isn't starting up because of a bad line in
wokd.conf:
2016/09/25 10:36:31 [emerg] 1742#1742: "user" directive is not allowed here
in /etc/nginx/conf.d/wok.conf:25

That's these lines:
user  www-data;
worker_processes  1;

**Comment by philled** [comment]
On 25 September 2016 at 10:40, Phill Edwards <philledwards@gmail.com> wrote:

> BTW, after encountering the issue above I then rebooted the Kimchi VM and
>> thyings are now in a worse state because when I go to
>> https://192.168.0.109:8001/ I get an error message in the browser saying
>> "502 Bad Gateway     nginx/1.10.0 (Ubuntu)".
>>
>
> More info... it looks like nginx isn't starting up because of a bad line
> in wokd.conf:
> 2016/09/25 10:36:31 [emerg] 1742#1742: "user" directive is not allowed
> here in /etc/nginx/conf.d/wok.conf:25
>
> That's these lines:
> user  www-data;
> worker_processes  1;
>
>
From what I've read there's an issue where wokd can't run if nginx is
already running. I've disabled nginx running on boot but I'm still getting
issues.

$ sudo systemctl status wokd
● wokd.service - Wok - Webserver Originated from Kimchi
   Loaded: loaded (/lib/systemd/system/wokd.service; enabled; vendor
preset: enabled)
  Drop-In: /etc/systemd/system/wokd.service.d
           └─kimchi.conf
   Active: active (running) since Sun 2016-09-25 11:40:28 AEST; 3s ago
  Process: 1487 ExecStop=/bin/kill -TERM $MAINPID (code=exited,
status=1/FAILURE)
 Main PID: 1526 (python2)
    Tasks: 6
   Memory: 76.5M
      CPU: 435ms
   CGroup: /system.slice/wokd.service
           ├─1380 nginx: master process nginx -c /etc/nginx/conf.d/wok.con
           ├─1381 nginx: worker process
           └─1526 python2 /usr/bin/wokd

Sep 25 11:40:29 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8001
failed (98: Address already in use)
Sep 25 11:40:29 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8000
failed (98: Address already in use)
Sep 25 11:40:29 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8001
failed (98: Address already in use)
Sep 25 11:40:29 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8000
failed (98: Address already in use)
Sep 25 11:40:30 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8001
failed (98: Address already in use)
Sep 25 11:40:30 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8000
failed (98: Address already in use)
Sep 25 11:40:30 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8001
failed (98: Address already in use)
Sep 25 11:40:30 kim wokd[1526]: nginx: [emerg] bind() to 0.0.0.0:8000
failed (98: Address already in use)
Sep 25 11:40:31 kim wokd[1526]: nginx: [emerg] still could not bind()
Sep 25 11:40:31 kim wokd[1526]: libvirt: XML-RPC error : Cannot recv data:
Host key verification failed.: Connection reset by peer

So it looks like there's some contention over ports 8000 and 8001 but how
can that be if ngninx isn't running at boot?

It also looks like there's a problem connecting to the KVM host. I've set
up password-less SSH and installed ssh-askpass. I've also tested from virsh
on the Kimchi box and that connects to the KVM host just fine so I don't
understand what the problem here is either.

If anyone could give me some pointers that would be great. This definitely
is not a straightforward install > configure > run exercise. I'm keeping
instructions which I'll share at the end if I can get this working because
the ones on the web site don't provide enough information.

**Comment by alinefm** [comment]
BTW, if you need quick answers it is better to ping me at #kimchi on OFTC.

**Comment by alinefm** [comment]
Also the error "502 Bad Gateway" happens when nginx is up but Wok is not completely up yet.
If you refresh the page time by time, you should get the Wok login page.

**Comment by CristianSP** [comment]
Hi all,

I'm following a similar path, to setup a VM with Kimchi which should then connect to the underline KVM physical host, all based on the CentOS distribution. Unfortunately, in the end no solution has worked out (virtualization tab is not listed in the UI) and from what I've seen the Kimchi plugin import fails because of a background SSH to the KVM host:
```bash
ssh -l root [host] sh -c 'if 'nc' -q 2>&1 | grep "requires an argument" >/dev/null 2>&1; then ARG=-q0;else ARG=;fi;'nc' $ARG -U /var/run/libvirt/libvirt-sock'
```

Steps to replicate:
1. Install and setup KVM physical host using minimal ISO.
2. Create a new VM for Kimchi
3. Add all necessary deps and packages on the VM
```bash
yum -y install acpid deltarpm epel-release bash-completion
yum -y update
reboot
yum install http://kimchi-project.github.io/wok/downloads/wok-2.3.0-0.el7.centos.noarch.rpm http://kimchi-project.github.io/gingerbase/downloads/ginger-base-2.2.1-0.el7.centos.noarch.rpm http://kimchi-project.github.io/kimchi/downloads/kimchi-2.3.0-0.el7.centos.noarch.rpm
yum install python-pip
pip install --upgrade pip
pip install numpy
```

4. Configure SSH passwordless between VM guest and KVM physical host
```bash
ssh-keygen -t rsa -N '' -f ~/.ssh/id_rsa
ssh-copy-id -i ~/.ssh/id_rsa.pub root@[host]
```

5. Check if connection works
```bash
virsh -c qemu+ssh://root@[host]/system nodeinfo
```

6. Configure libvirt 
```bash
sed -i 's|#uri_default =.*|uri_default = "qemu+ssh://root@[host]/system"|' /etc/libvirt/libvirt.conf
```

7. Temporarily disable any security on it
```bash
setenforce 0
systemctl stop firewalld
```

8. Launch wok/kimchi
```bash
systemctl daemon-reload
systemctl restart wokd.service
```

9. Output of running **wokd --env=development** with new URI [bad.txt](https://github.com/kimchi-project/kimchi/files/547152/bad.txt)
10. Output of running **wokd --env=development** with default URI 
[good.txt](https://github.com/kimchi-project/kimchi/files/547154/good.txt)

If I replicate the steps on the physical node the plugin is successfully loaded and seen in the UI.

**Comment by philled** [comment]
>
> BTW, if you need quick answers it is better to ping me at #kimchi on OFTC.
>
Hi Aline, thanks but I don't know what OFTC is. Can you please be a bit
more specific?

**Comment by philled** [comment]
>
> Let step back a bit and let me tell you what I did:
>
>    1. Fresh Ubuntu 16.04 install in a KVM guest
>    2. sudo apt-get update
>    3. sudo apt-get upgrade (to make sure everything is on latest version)
>    4. Download Wok (http://kimchi-project.github.
>    io/wok/downloads/wok-2.3.0-0.noarch.deb
>    <http://kimchi-project.github.io/wok/downloads/wok-2.3.0-0.noarch.deb>),
>    Ginger Base (http://kimchi-project.github.io/gingerbase/downloads/
>    ginger-base-2.2.1-0.noarch.deb) and Kimchi (
>    http://kimchi-project.github.io/kimchi/downloads/kimchi-2.
>    3.0-0.noarch.deb
>    <http://kimchi-project.github.io/kimchi/downloads/kimchi-2.3.0-0.noarch.deb>
>    )
>    5. sudo dpkg -i wok.deb gingerbase.deb kimchi.deb
>    6. The above command WILL FAIL!! As my system does not have all the
>    dependencies needed already installed. SO:
>    7. sudo apt-get install -f (it will install all the dependencies
>    needed + wok.deb + gingerbase.deb + kimchi.deb)
>    8. Open 8000, 8001 on firewall
>    9. sudo service wokd start
>
> And it is all I needed to do.
>
> What did you do different than me?
>
Hi Aline, thanks for your help on this. I did all the things you did above.
I wil point out again, though, that the machine I'm installing Kimchi on is
a virtual machine, running on the same KVM host that it's supposed to
manage - is that the same for you?

We should probably abandon this thread and revert to the other thread where
you're helping me through a libvirt connection issue which I'm hoping is
the final obstacle to overcome.

Regards,
Phill

**Comment by CristianSP** [comment]
Hi Phill,

Hope you did not mind taking over but I felt our problems are similar, we both want a dedicated VM guest of kimchi to control the underline physical host (bare-metal), which now I don't think it's possible but I'll let the Aline to confirm this or not.

What convinced me was that even qemu+tcp w/o any authentication enabled is not able to bring up kimchi plugin. Thou the bash command to test the connection works:
```bash
### On the physical host
# enable listening
vi /etc/sysconfig/libvirtd
# enable tcp and disable all possible authentications 
vi /etc/libvirt/libvirtd.conf
# reload services
systemctl daemon-reload
systemctl restart libvirtd

# From the guest VM
virsh -c qemu+tcp://++host-ip or fqdn++/system nodeinfo
```

So it has nothing to do with user rights (ssh passwordless) or nested virtualization (that's needed only if you want to have a KVM enabled in a KVM guest - not the case).

Hope it helps.
Cristi.

**Comment by alinefm** [comment]
>
> BTW, if you need quick answers it is better to ping me at #kimchi on OFTC.
>
> Hi Aline, thanks but I don't know what OFTC is. Can you please be a bit
> more specific?

OFTC is a IRC server (irc.oftc.net) and there is a #kimchi channel there.

**Comment by alinefm** [comment]
Ok. So what you are doing is different than me.
My Kimchi VM is running with the default libvirt settings, ie, it is not pointing to an external libvirt URI (as you are doing).

Kimchi uses the following code to connect to the libvirt service:
```
import libvirt
conn = libvirt.open(None)
```
The "None" parameter means to connect to the running libvirt instead of any special URI.
So you will need to have the libvirt service proper pointing to the external URI and up/running.

You can verify it directly through libvirt by doing:
```
import libvirt
conn = libvirt.open(None)
print conn.listAllDomains()
```
Is that printing all the VMs there are in the connected system?

Also I'd enable nested virtualization on VM.
Check https://kashyapc.com/2012/01/14/nested-virtualization-with-kvm-intel/ for more details.

Answering your doubts: Yeap! Kimchi is supposed to connect to the libvirt service running in the same machine Kimchi is installed - no matters libvirt is connected locally or using an external URI.

If you are able to access Wok and there is no Virtualization tab, that means the Kimchi fails to load and there would be an error in the logs.

My suggestion is to stop using the wokd service and run wokd manually to get more output details:

```
service wokd stop
wokd --environment=dev
```

I will try to setup the same environment as yours and input the results I find here.

**Comment by philled** [comment]
Hi Cristi, thanks for the info. You may be correct - but hopefully not! I
thought others had got this working, but may have misunderstood their setup.

Aline, it would be great if you could comment here.

I think one of the issues with Kimchi is that it seems to be a small user
group. That coupled with the documentation, which is unfortunately quite
sparse, makes it very difficult to resolve issues.

Regards,
Phill

On 26 Oct 2016 5:16 PM, "Cristian SABAILA" <notifications@github.com> wrote:

Hi Phill,

Hope you did not mind taking over but I felt our problems are similar, we
both want a dedicated VM guest of kimchi to control the underline physical
host (bare-metal), which now I don't think it's possible but I'll let the
Aline to confirm this or not.

What convinced me was that even qemu+tcp w/o any authentication enabled is
not able to bring up kimchi plugin. Thou the bash command to test the
connection works:

### On the physical host# enable listening
vi /etc/sysconfig/libvirtd# enable tcp and disable all possible authentications
vi /etc/libvirt/libvirtd.conf# reload services
systemctl daemon-reload
systemctl restart libvirtd
# From the guest VM
virsh -c qemu+tcp://++host-ip or fqdn++/system nodeinfo

So it has nothing to do with user rights (ssh passwordless) or nested
virtualization (that's needed only if you want to have a KVM enabled in a
KVM guest - not the case).

Hope it helps.
Cristi.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kimchi-project/kimchi/issues/952#issuecomment-256258955>,
or mute the thread

**Comment by philled** [comment]
>
> You can verify it directly through libvirt by doing:
>
> import libvirt
> conn = libvirt.open(None)
> print conn.listAllDomains()
>
> Is that printing all the VMs there are in the connected system?
>
Om the Kimchi VM:

#!/usr/bin/python
import libvirt
conn = libvirt.open(None)
print conn.listAllDomains()

returns this:
[]


When I run it on the KVM host it returns this:
[<libvirt.virDomain object at 0x7f8840754710>, <libvirt.virDomain object at
0x7f8840754750>, <libvirt.virDomain object at 0x7f8840754790>,
<libvirt.virDomain object at 0x7f88407547d0>, <libvirt.virDomain object at
0x7f8840754810>, <libvirt.virDomain object at 0x7f8840754850>,
<libvirt.virDomain object at 0x7f8840754890>, <libvirt.virDomain object at
0x7f88407548d0>, <libvirt.virDomain object at 0x7f8840754910>]

Also I'd enable nested virtualization on VM.
> Check https://kashyapc.com/2012/01/14/nested-virtualization-with-
> kvm-intel/ for more details.
>
I had actually looked at this post one before but I didn't fully understand
it. I don't configure and run my VMs this way. However, from what I could
understand the iumportant thing is to make sure virtualisation is enabled
for the VM's virtual CPU - and it is in my Kimchi VM:

$ grep -E "(vmx|svm)" --color=always /proc/cpuinfo
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca
cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx rdtscp lm
constant_tsc rep_good nopl eagerfpu pni pclmulqdq vmx ssse3 cx16 pcid
sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor
lahf_lm tpr_shadow vnmi flexpriority ept vpid xsaveopt arat


> Answering your doubts: Yeap! Kimchi is supposed to connect to the libvirt
> service running in the same machine Kimchi is installed - no matters
> libvirt is connected locally or using an external URI.
>
> If you are able to access Wok and there is no Virtualization tab, that
> means the Kimchi fails to load and there would be an error in the logs.
>
> My suggestion is to stop using the wokd service and run wokd manually to
> get more output details:
>
> service wokd stop
> wokd --environment=dev
>
>  OK, this is what I see when I run these commands. Does this help narriow
down the problem?

$ sudo wokd --environment=dev
The authenticity of host '192.168.0.102 (192.168.0.102)' can't be
established.
ECDSA key fingerprint is SHA256:JQ+HBntIz8P3wm1FA2i1koitlIxNgSeIbRE/Ezy0JVk.
Are you sure you want to continue connecting (yes/no)? yes
phill@192.168.0.102's password:
libvirt: QEMU Driver error : Domain not found: no domain with matching name
'FEATURETEST_VM'
libvirt: Storage Driver error : Storage pool not found: no storage pool
with matching name 'FEATURETEST_POOL'

*** Kimchi: Running feature tests ***
Service Libvirtd running ...: True
libvirt:  error : internal error: Child process (/sbin/showmount
--no-headers --exports 127.0.0.1) unexpected exit status 127: libvirt:
 error : cannot execute binary /sbin/showmount: No such file or directory
internal error: Child process (/sbin/showmount --no-headers --exports
127.0.0.1) unexpected exit status 127: libvirt:  error : cannot execute
binary /sbin/showmount: No such file or directory

NFS Target Probe support ...: True
Fibre Channel Host support .: True
Kernel VFIO support ........: True
Network Manager running ....: False
Memory Hotplug support .....: True
*** Kimchi: Feature tests completed ***

libvirt: Storage Driver error : Storage pool not found: no storage pool
with matching name 'ISO'
phill@192.168.0.102's password:
Failed to import plugin plugins.kimchi.Kimchi, error:
*** Ginger Base: Running capabilities tests ***
System Report Tool ...: True
Loading AptUpdate features.
System Update Tool ...: True
Repo Management Tool .: deb
*** Ginger Base: Capabilities tests completed ***
[27/Oct/2016:19:53:44] ENGINE Listening for SIGHUP.
[27/Oct/2016:19:53:44] ENGINE Listening for SIGTERM.
[27/Oct/2016:19:53:44] ENGINE Listening for SIGUSR1.
[27/Oct/2016:19:53:44] ENGINE Bus STARTING
[27/Oct/2016:19:53:44] ENGINE Started monitor thread 'Autoreloader'.
[27/Oct/2016:19:53:44] ENGINE Serving on http://127.0.0.1:8010

*** Kimchi: Running dependable feature tests ***
[27/Oct/2016:19:53:44] ENGINE Started monitor thread 'Session cleanup'.
127.0.0.1 - - [27/Oct/2016:19:53:44] "HEAD
/plugins/kimchi/images/icon-fedora.png HTTP/1.1" 404 1597 "" ""
QEMU stream support .......: False
Libvirt Stream Protocols ..: ['http', 'https', 'ftp', 'ftps', 'tftp']
*** Kimchi: Dependable feature tests completed ***

[27/Oct/2016:19:53:44] ENGINE Bus STARTED

I still don't see any virtual machine information in the web interface
after starting Wok this way.

**Comment by alinefm** [comment]
According to your logs, Kimchi failed to be loaded:

```
Failed to import plugin plugins.kimchi.Kimchi, error:
```

Hack server.py, line 197 to get the right reason of the failure. I am not sure why it is not being printed out.

**Comment by erpalma** [comment]
The problem is that the env variable LIBVIRT_DEFAULT_URI has priority over libvirt.conf and by default is set to LIBVIRT_DEFAULT_URI=qemu:///system

I've forced the env variable but kimchi does not appear in wok. 
`LIBVIRT_DEFAULT_URI="qemu+ssh://root@10.50.0.1/system" wokd --environment=dev
*** Ginger Base: Running capabilities tests ***
System Report Tool ...: True
Loading AptUpdate features.
System Update Tool ...: True
Repo Management Tool .: deb
*** Ginger Base: Capabilities tests completed ***
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'FEATURETEST_VM'
libvirt: Storage Driver error : Storage pool not found: no storage pool with matching name 'FEATURETEST_POOL'

*** Kimchi: Running feature tests ***
Service Libvirtd running ...: True
libvirt:  error : internal error: Child process (/sbin/showmount --no-headers --exports 127.0.0.1) unexpected exit status 127: libvirt:  error : cannot execute binary /sbin/showmount: No such file or directory
internal error: Child process (/sbin/showmount --no-headers --exports 127.0.0.1) unexpected exit status 127: libvirt:  error : cannot execute binary /sbin/showmount: No such file or directory

NFS Target Probe support ...: True
Fibre Channel Host support .: True
Kernel VFIO support ........: True
Network Manager running ....: False
Memory Hotplug support .....: True
*** Kimchi: Feature tests completed ***

Failed to import plugin plugins.kimchi.Kimchi, error:
[06/Dec/2016:16:34:12] ENGINE Listening for SIGHUP.
[06/Dec/2016:16:34:12] ENGINE Listening for SIGTERM.
[06/Dec/2016:16:34:12] ENGINE Listening for SIGUSR1.
[06/Dec/2016:16:34:12] ENGINE Bus STARTING
[06/Dec/2016:16:34:12] ENGINE Started monitor thread 'Autoreloader'.
[06/Dec/2016:16:34:12] ENGINE Serving on http://127.0.0.1:8010

*** Kimchi: Running dependable feature tests ***
[06/Dec/2016:16:34:12] ENGINE Started monitor thread 'Session cleanup'.
127.0.0.1 - - [06/Dec/2016:16:34:12] "HEAD /plugins/kimchi/images/icon-fedora.png HTTP/1.1" 404 1597 "" ""
QEMU stream support .......: False
Libvirt Stream Protocols ..: ['http', 'https', 'ftp', 'ftps', 'tftp']
*** Kimchi: Dependable feature tests completed ***

[06/Dec/2016:16:34:12] ENGINE Bus STARTED`

**Comment by AntiPaste** [comment]
Hi, bringing this back to life. I believe I'm in the same situation as where @philled left off.

After hacking server.py, the real exception:
```plaintext
[Errno 2] No such file or directory: '/var/run/libvirt/qemu/KVMUSERTEST_VM.pid'
```

This apparently originates here in user probing: https://github.com/kimchi-project/kimchi/blob/master/kvmusertests.py#L65

The bug seems logical if I understand correctly: the user probing creates a guest and attempts to read the owner of the PID file, but when using a remote libvirt the PID file cannot be found in the filesystem that Kimchi is running in.

**Comment by amit213** [comment]
Fyi. I was able to get the kimchi working with nested virtualization. 

i.e. Bare metal node --> Ubuntu OS --> kvm guest qemu VM -->  kimchi infra + Qemu libvirtd ==> [ child ubuntu vm1, child centos vm2, child ubuntu vmN qcow2 ]

**Comment by iwanovich** [comment]
@amit213 I think that your situation is slightly than @philled's situation. In your case you can't manage the VM's of your (bare-metal) host. You run another hypervisor in a VM which can be managed by Kimchi. Am I correct?

The kvmusertest basically hinders the use case scenario of @philled. I guess In order to get this to work the code has to be adjusted to remove the kvmusertest or its dependency to look for a pid on Kimchi's system that is a side effect of creating a VM.

In short: In the current state, Kimchi doesn't support managing the host.
Please correct me if I'm wrong.
